### ISSUE TITLE
Implement `process.stdin.setRawMode` (and `ReadlineInterface.setRawMode`)

### ISSUE THREAD

**Issue by ThatOneBro** [oven-sh/bun#2025]
Need to implement `setRawMode` on `process.stdin` and match Node behavior around "raw mode" and "normal mode". 

This is nearly finished but I still need to add cleanup, and potentially add bindings for Zig, so that Zig native streams can control raw mode (only if necessary).

**Comment by vsnthdev** [comment]
Egerly waiting for this, so I can implement cool terminal prompts for Bun 👏

**Comment by AlexBlokh** [comment]
we're aiming to migrate `drizzle-kit` to bun and that's currently a blocker
I'm not sure there's a way we can help with the PR, we're no good with neither C nor Zig

**Comment by Jarred-Sumner** [comment]
We are aiming to ship this in Bun v0.6.16 or Bun v0.7 (whichever comes first)

**Comment by AlexBlokh** [comment]
sounds amazing!
It all comes down to the shittiest thing in JS - esm + cjs. We've spent massive amount of time with esbuild, esbuild-register, esbuild-kit/esm(cjs)-loader, custom node loader, swc, whatever. It just doesn't work and seems like what you've done with bun might just be the only option we've left with, bun is really the way it should've been done

**Comment by fabiospampinato** [comment]
Worth checking if this is fixed once raw mode is implemented (currently it outputs two lines rather than 1 like in Node, and it also keeps the process alive for some reason).

```ts
import prompts from 'prompts';

await prompts ({
  type: 'text',
  name: 'asd',
  initial: 'John Doe',
  message: 'What is your name?',
})
```

**Comment by gtrabanco** [comment]
For me still does not work as expected with this code:

```ts
const { createInterface } = require('node:readline/promises');

async function ynInKey(questionPrompt = null, defaultAnswer = true, abortTime = 10_000) {
  const signal = AbortSignal.timeout(abortTime);
  const readline = createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const options = defaultAnswer ? '[Y/n]' : '[y/N]';
  const prompt = `${questionPrompt} ${options} ${readline.getPrompt()}`;

  readline.question(prompt, { signal });
  return new Promise((resolve, reject) => {
    process.stdin.setRawMode(true);
    process.stdin.resume();
    process.stdin.on('data', (data) => {
      process.stdin.pause();
      console.log();
      const key = data.toString('utf-8');
      process.stdin.setRawMode(false);
      readline.close();
      if (key.toLowerCase() === 'y' || key.toLowerCase() !== 'n' && defaultAnswer) return resolve();

      return reject(new Error('No'));
    });
  });
}

await ynInKey('Do you want to continue?', true, 10_000);
```

I am still have to press `enter` key twice instead of detect `y` or `n` char.

**Comment by vsnthdev** [comment]
I can confirm this issue has been resolved in Bun `v0.8.0` ✅ 

Shall we close this issue now?
